File: C/11_data_heap_red_black_tree.c
/*
 * 任务：堆、红黑树。
 * 编译：gcc -std=c11 -O2 -Wall -Wextra 11_11_data_heap_red_black_tree.c
 */

#include <stdio.h>
#include <stdlib.h>


/* ---------- 二叉堆（数组实现，小根堆） ---------- */

#define HEAP_CAP 100

typedef struct {
    int data[HEAP_CAP];
    int n;
} MinHeap;

static void heap_init(MinHeap *h) { h->n = 0; }

static void heap_swap(int *a, int *b) { int t = *a; *a = *b; *b = t; }

static int heap_push(MinHeap *h, int x) {
    if (h->n >= HEAP_CAP) return 0;
    int i = h->n++;
    h->data[i] = x;
    while (i > 0) {
        int p = (i - 1) / 2;
        if (h->data[p] <= h->data[i]) break;
        heap_swap(&h->data[p], &h->data[i]);
        i = p;
    }
    return 1;
}

static int heap_pop(MinHeap *h, int *out) {
    if (h->n == 0) return 0;
    *out = h->data[0];
    h->data[0] = h->data[--h->n];
    int i = 0;
    for (;;) {
        int l = 2 * i + 1, r = 2 * i + 2, m = i;
        if (l < h->n && h->data[l] < h->data[m]) m = l;
        if (r < h->n && h->data[r] < h->data[m]) m = r;
        if (m == i) break;
        heap_swap(&h->data[i], &h->data[m]);
        i = m;
    }
    return 1;
}

static void heap_demo(void) {
    MinHeap h;
    int x;
    int vals[] = {5, 3, 8, 1, 9, 2, 7};
    heap_init(&h);
    for (int i = 0; i < (int)(sizeof vals / sizeof vals[0]); ++i) heap_push(&h, vals[i]);

    printf("heap pop order:");
    while (heap_pop(&h, &x)) printf(" %d", x);
    printf("\n");
}


/* ---------- 红黑树（左倾红黑树 LLRB） ---------- */

typedef enum { RED, BLACK } Color;

typedef struct RBNode {
    int key;
    Color color;
    struct RBNode *left, *right;
} RBNode;

static int is_red(const RBNode *h) { return h != NULL && h->color == RED; }

static RBNode *rb_new(int key, Color color) {
    RBNode *p = malloc(sizeof *p);
    if (!p) { perror("malloc"); exit(1); }
    p->key = key;
    p->color = color;
    p->left = p->right = NULL;
    return p;
}

static RBNode *rotate_left(RBNode *h) {
    RBNode *x = h->right;
    h->right = x->left;
    x->left = h;
    x->color = h->color;
    h->color = RED;
    return x;
}

static RBNode *rotate_right(RBNode *h) {
    RBNode *x = h->left;
    h->left = x->right;
    x->right = h;
    x->color = h->color;
    h->color = RED;
    return x;
}

static void flip_colors(RBNode *h) {
    h->color = RED;
    h->left->color = BLACK;
    h->right->color = BLACK;
}

static RBNode *rb_insert(RBNode *h, int key) {
    if (!h) return rb_new(key, RED);

    if (key < h->key)       h->left  = rb_insert(h->left, key);
    else if (key > h->key)  h->right = rb_insert(h->right, key);
    else                    return h;

    if (is_red(h->right) && !is_red(h->left))          h = rotate_left(h);
    if (is_red(h->left) && is_red(h->left->left))      h = rotate_right(h);
    if (is_red(h->left) && is_red(h->right))           flip_colors(h);
    return h;
}

static void rb_inorder(const RBNode *h) {
    if (!h) return;
    rb_inorder(h->left);
    printf("%d ", h->key);
    rb_inorder(h->right);
}

static void rb_free(RBNode *h) {
    if (!h) return;
    rb_free(h->left);
    rb_free(h->right);
    free(h);
}

static void rb_demo(void) {
    int vals[] = {7, 3, 18, 10, 22, 8, 11, 26, 2, 6, 13};
    RBNode *root = NULL;
    for (int i = 0; i < (int)(sizeof vals / sizeof vals[0]); ++i)
        root = rb_insert(root, vals[i]);
    root->color = BLACK;

    printf("red-black tree inorder:");
    rb_inorder(root);
    printf("\n");
    rb_free(root);
}


int main(void) {
    heap_demo();
    printf("\n");
    rb_demo();
    return 0;
}

File: Haskell/11_data_heap_red_black_tree.hs
-- 任务：堆、红黑树。
-- 运行：runghc 11_11_data_heap_red_black_tree.hs

import Data.List (foldl')
import Text.Printf (printf)


-- ---------- 二叉堆（列表实现，小根堆） ----------

-- 完全二叉树按层序存放在列表中：i 的孩子是 2i+1 和 2i+2。
pushHeap :: Ord a => a -> [a] -> [a]
pushHeap x h = siftUp (h ++ [x]) (length h)
  where
    siftUp arr i
        | i <= 0 = arr
        | arr !! p <= arr !! i = arr
        | otherwise = siftUp (swapAt i p arr) p
      where
        p = (i - 1) `div` 2

popHeap :: Ord a => [a] -> Maybe (a, [a])
popHeap [] = Nothing
popHeap h@(x : _)
    | n == 1 = Just (x, [])
    | otherwise = Just (x, siftDown (replaceHead (init h) (last h)) 0)
  where
    n = length h

    replaceHead (_ : xs) v = v : xs
    replaceHead [] _ = []

    siftDown arr i =
        let n' = length arr
            l = 2 * i + 1
            r = 2 * i + 2
            m = pickSmallest arr i l r
        in if m == i then arr else siftDown (swapAt i m arr) m

    pickSmallest arr i l r =
        let m1 = if l < length arr && arr !! l < arr !! i then l else i
        in if r < length arr && arr !! r < arr !! m1 then r else m1

swapAt :: Int -> Int -> [a] -> [a]
swapAt i j xs =
    [ if k == i then xs !! j
      else if k == j then xs !! i
      else xs !! k
    | k <- [0 .. length xs - 1]
    ]

heapFromList :: Ord a => [a] -> [a]
heapFromList = foldl' (flip pushHeap) []


-- ---------- 红黑树（左倾红黑树 LLRB） ----------

data Color = Red | Black
    deriving (Eq, Show)

data RBTree a = RBE | RBN Color (RBTree a) a (RBTree a)
    deriving (Show)

rbEmpty :: RBTree a
rbEmpty = RBE

rbInsert :: Ord a => a -> RBTree a -> RBTree a
rbInsert x t = blacken (ins t)
  where
    ins RBE = RBN Red RBE x RBE
    ins (RBN c l y r)
        | x < y     = fixup (RBN c (ins l) y r)
        | x > y     = fixup (RBN c l y (ins r))
        | otherwise = RBN c l y r

    fixup t1 =
        let t2 = if isRed (rightOf t1) && not (isRed (leftOf t1))
                    then rotateLeft t1
                    else t1
            t3 = if isRed (leftOf t2) && isRed (leftOf (leftOf t2))
                    then rotateRight t2
                    else t2
        in if isRed (leftOf t3) && isRed (rightOf t3)
              then flipColors t3
              else t3

    rotateLeft (RBN c l x r@(RBN _ rl y rr)) = RBN c (RBN Red l x rl) y rr
    rotateLeft t = t

    rotateRight (RBN c l@(RBN _ ll x lr) y r) = RBN c ll x (RBN Red lr y r)
    rotateRight t = t

    flipColors (RBN _ l x r) = RBN Red (setBlack l) x (setBlack r)
    flipColors t = t

    setBlack (RBN _ l x r) = RBN Black l x r
    setBlack t = t

    blacken (RBN _ l x r) = RBN Black l x r
    blacken t = t

isRed :: RBTree a -> Bool
isRed (RBN Red _ _ _) = True
isRed _ = False

leftOf :: RBTree a -> RBTree a
leftOf (RBN _ l _ _) = l
leftOf _ = RBE

rightOf :: RBTree a -> RBTree a
rightOf (RBN _ _ _ r) = r
rightOf _ = RBE

rbFromList :: Ord a => [a] -> RBTree a
rbFromList = foldl' (flip rbInsert) RBE

rbInorder :: RBTree a -> [a]
rbInorder RBE = []
rbInorder (RBN _ l x r) = rbInorder l ++ [x] ++ rbInorder r


main :: IO ()
main = do
    let h = heapFromList [5, 3, 8, 1, 9, 2, 7] :: [Int]
    let popAll heap acc = case popHeap heap of
                              Nothing -> reverse acc
                              Just (x, rest) -> popAll rest (x : acc)
    printf "heap pop order: %s\n" (show (popAll h []))

    putStrLn ""
    let tree = rbFromList [7, 3, 18, 10, 22, 8, 11, 26, 2, 6, 13] :: RBTree Int
    printf "red-black tree inorder: %s\n" (show (rbInorder tree))